qutip/core/qobj.py

```diff
@@ -296,58 +271,110 @@ class Qobj:
 
     def __init__(self, arg=None, dims=None, type=None,
                  copy=True, superrep=None, isherm=None, isunitary=None):
-        self._initialize_data(arg, dims, copy)
-        self.type = type or type_from_dims(self.dims)
         self._isherm = isherm
         self._isunitary = isunitary
+        self._superrep = None
+        if isinstance(dims, list):
+            dims = Dimensions(dims)
+        self._initialize_data(arg, dims, copy)
 
-        if self.type == 'super' and type_from_dims(self.dims) == 'oper':
-            if self._data.shape[0] != self._data.shape[1]:
-                raise ValueError("".join([
-                    "cannot build superoperator from nonsquare data of shape ",
-                    repr(self._data.shape),
-                ]))
-            root = int(np.sqrt(self._data.shape[0]))
-            if root * root != self._data.shape[0]:
-                raise ValueError("".join([
-                    "cannot build superoperator from nonsquare subspaces ",
-                    "of size ",
-                    repr(self._data.shape[0]),
-                ]))
-            self.dims = [[[root]]*2]*2
-        if self.type in ['super', 'operator-ket', 'operator-bra']:
-            superrep = superrep or 'super'
+        # Dims are guessed from the data and need to be changed to super.
+        if (
+            type in ['super', 'operator-ket', 'operator-bra']
+            and self._dims.type in ['oper', 'ket', 'bra']
+        ):
+            root_right = int(np.sqrt(self._data.shape[0]))
+            root_left = int(np.sqrt(self._data.shape[1]))
+            if (
+                root_right * root_right != self._data.shape[0]
+                and root_left * root_left != self._data.shape[1]
+            ):
+                raise ValueError(
+                    "cannot build superoperator from nonsquare subspaces"
+                )
+            self.dims = [[[root_right]]*2, [[root_left]]*2]
+
+        self.type = type
         self.superrep = superrep
 
     def copy(self):
         """Create identical copy"""
         return Qobj(arg=self._data,
                     dims=self.dims,
-                    type=self.type,
-                    superrep=self.superrep,
                     isherm=self._isherm,
                     isunitary=self._isunitary,
                     copy=True)
 
+    @property
+    def dims(self):
+        return self._dims.as_list()
+
+    @dims.setter
+    def dims(self, dims):
+        dims = Dimensions(dims, rep=self.superrep)
+        if self._data and dims.shape != self._data.shape:
+            raise ValueError('Provided dimensions do not match the data: ' +
+                             f"{dims.shape} vs {self._data.shape}")
+        self._dims = dims
+        self.type = self._dims.type
+
+    @property
+    def type(self):
+        return self._type
+
+    @type.setter
+    def type(self, val):
+        if not val:
```

Comment:
It's not clear to me what the use case for setting type is.

Reply:
For 1x1 Qobj, they can be anything...
Also for security / backward compatibility.
In v4 there is nothing stopping you from doing:
```
qobj = Qobj(data)
qobj.type = "abc"
```
Now it would at least raise error when it does not fit.